lib/src/modules/task.cc

```diff
@@ -240,30 +239,43 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
 //    the same task file.
 // The downloaded task file's permissions will be set to rwx for user and rx for
 // group for non-Windows OSes.
-static void downloadTaskFile(const std::vector<std::string>& master_uris,
+static bool downloadTaskFile(const std::vector<std::string>& master_uris,
                              lth_curl::client& client,
                              const fs::path& file_path,
                              const lth_jc::JsonContainer& uri) {
-    auto url = master_uris[0] + createUrlEndpoint(uri);
-    lth_curl::request req(url);
-    // timeout from connection after one minute, can configure
-    req.connection_timeout(60000);
+    auto endpoint = createUrlEndpoint(uri);
+    for (auto& master_uri : master_uris) {
+        auto url = master_uri + endpoint;
+        lth_curl::request req(url);
+        // timeout from connection after one minute, can configure
+        req.connection_timeout(60000);
+        try {
 #ifndef _WIN32
-    client.download_file(req, file_path.string(), NIX_TASK_FILE_PERMS);
+            client.download_file(req, file_path.string(), NIX_TASK_FILE_PERMS);
 #else
-    client.download_file(req, file_path.string());
+            client.download_file(req, file_path.string());
 #endif
+            break;
+        } catch (lth_curl::http_file_download_exception& e) {
+            // Do nothing here -- we want to try the next master-uri.
+
+            // TODO: Should we collect all the exception messages here to pass in to the overall
+            // Module::ProcessingError if all master_uris fail? Might not be a bad idea.
```

Comment:
Okay, so I'm thinking something along the following lines for the Leatherman changes:

(1) If a filesystem_error is thrown, then that is a file operation failure (writing to disk)

(2) If curl_easy_perform returns CURL_WRITE_ERROR, that is also a file operation failure.

(3) Otherwise, it's a server error.

It seems like (1) and (2) require a new exception class representing file operation failures, unless we want to use fs::filesystem_error. (3) is an http_file_download_exception.